pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java

```diff
@@ -58,6 +63,12 @@ public TopicTransactionBuffer(PersistentTopic topic) {
         topic.getManagedLedger().asyncAddEntry(buffer, new AsyncCallbacks.AddEntryCallback() {
             @Override
             public void addComplete(Position position, Object ctx) {
+                if (!ongoingTxns.containsKey(txnId)) {
+                    ongoingTxns.put(txnId, (PositionImpl) position);
+                    PositionImpl firstPosition = ongoingTxns.get(ongoingTxns.firstKey());
```

Comment:
The LinkedMap will throw NoSuchElementException while the map is empty. And I think the LinkedMap is not a thread-safe container?

Reply:
Yes, it not a thread-safe container, i lock all logical what change this map.